File: 2024-dgu-spring-camp/13/4666.cpp
// 수 정렬하기
// https://study.helloalgo.co.kr/study/1019/room/600/6576

// python 시간 초과로 cpp로 통과.

#include <iostream>
#include <vector>
using namespace std;

vector<int> heap;

int getLeftIdx(int cur)
{
    return cur * 2 + 1;
}
int getRightIdx(int cur)
{
    return cur * 2 + 2;
}
int getParentIdx(int cur)
{
    return cur % 2 == 0 ? cur / 2 - 1 : cur / 2;
}

int swap(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

void heapUp(int idx)
{
    int pIdx = getParentIdx(idx);
    while (pIdx >= 0)
    {
        if (heap[idx] < heap[pIdx])
        {
            swap(heap[idx], heap[pIdx]);
        }

        idx = pIdx;
        pIdx = getParentIdx(idx);
    }
}

void heapDown(int idx)
{
    while (true)
    {
        int left, right;
        left = getLeftIdx(idx);
        right = getRightIdx(idx);

        if (left < heap.size() && heap[left] < heap[idx] && (right >= heap.size() || heap[left] < heap[right]))
        {
            swap(heap[idx], heap[left]);
            idx = left;
        }
        else if (right < heap.size() && heap[right] < heap[idx] && heap[right] <= heap[left])
        {
            swap(heap[idx], heap[right]);
            idx = right;
        }
        else
        {
            break;
        }
    }
}

void add(int num)
{
    heap.push_back(num);
    heapUp(heap.size() - 1);
}

int remove()
{
    if (heap.size() == 0)
    {
        int result = heap.back();
        heap.pop_back();
        return result;
    }

    int result = heap[0];
    heap[0] = heap.back();
    heap.pop_back();
    heapDown(0);

    return result;
}

int main()
{
    int N, temp;
    cin >> N;
    for (int i = 0; i < N; i++)
    {
        cin >> temp;
        add(temp);
    }

    while (heap.size() > 0)
    {
        cout << remove() << '\n';
    }

    return 0;
}

File: 2024-dgu-spring-camp/13/4666.py
# 수 정렬하기
# https://study.helloalgo.co.kr/study/1019/room/600/6576/source/829006/

# Array 트리 연습용 min-heap 직접 구현.
# - heap-sort

import sys

"""
        0
   1          2
 3   4     5     6
7 8 9 10 11 12 13 14

left: cur*2 + 1
right: cur*2 + 2
parent: cur//2 -1 if cur%2 == 0 else cur//2
"""

heap = []

def getLeftIdx(cur):
    return cur*2 + 1
def getRightIdx(cur):
    return cur*2 + 2
def getParentIdx(cur):
    return cur//2 - 1 if cur%2 == 0 else cur//2

def swap(arr, idx1, idx2):
    arr[idx1], arr[idx2] = arr[idx2], arr[idx1]

def heapUp(idx):
    pIdx = getParentIdx(idx)
    while pIdx >= 0:
        if heap[idx] < heap[pIdx]:
            swap(heap, idx, pIdx)
        
        idx = pIdx
        pIdx = getParentIdx(idx)
        
def heapDown(idx):
    while True:
        left, right = getLeftIdx(idx), getRightIdx(idx)
        if left < len(heap) and heap[left] < heap[idx] and (right >= len(heap) or heap[left] <= heap[right]):
            swap(heap, idx, left)
            idx = left
        elif right < len(heap) and heap[right] < heap[idx] and heap[right] <= heap[left]:
            swap(heap, idx, right)
            idx = right
        else:
            break

def add(num):
    heap.append(num)
    heapUp(len(heap)-1)
    
def remove():
    if len(heap) == 0:
        return heap.pop()
    
    result = heap[0]
    heap[0] = heap[-1]
    heap.pop()
    heapDown(0)
    
    return result

N = int(sys.stdin.readline())
for _ in range(N):
    x = int(sys.stdin.readline())
    add(x)

answers = []
while len(heap) > 0:
    answers.append(str(remove()))
    
print('\n'.join(answers))
